### PR TITLE
Dynamic Workers and Fault tolerance 

### PR DESCRIPTION
**Pull request by apoms**
This PR adds the ability to add/remove workers during job execution and provides basic fault tolerance.

The master supports the addition of new workers during the execution of an in-progress job. New workers will be able to request work just like any other worker. The master also supports removing workers during the job. If a worker is removed while executing a job, the master will reassign its tasks to other workers.

Currently we aren't polling workers to ensure they are alive and actively removing them from the list of active workers (and thus reassigning them). This should be done before this PR is fully merged (as it represents true fault tolerance). As of now, a worker (or other process) must explicitly call `UnregisterWorker` on the master to remove the worker.

### CODE REVIEW

File: python/scannerpy/database.py [hunk 1]
```diff
@@ -328,6 +349,85 @@ def _stop_heartbeat(self):
...
+            self._worker_conns = None
+            raise ScannerException(
+                'Timed out waiting for workers to connect to master')
+
+
+    def stop_worker(self, worker_address):
+        assert self._debug == False
+        # Find worker by address
+        i = 0
+        for addr in self._worker_addresses:
```
Comment (willcrichton) [inline python/scannerpy/database.py]:
```python
i = self._worker_addresses.index(addr)
```

File: python/scannerpy/database.py [hunk 2]
```diff
@@ -328,6 +349,85 @@ def _stop_heartbeat(self):
...
+            if addr == worker_address:
+                break
+            i += 1
+        assert i != len(self._worker_addresses)
+
+        self._worker_conns[i]
```
Comment (willcrichton) [inline python/scannerpy/database.py]: What's happening on these two lines?

File: scanner/engine/worker.cpp [hunk 3]
```diff
@@ -814,7 +806,27 @@ WorkerImpl::WorkerImpl(DatabaseParameters& db_params,
 }
 
 WorkerImpl::~WorkerImpl() {
+  State state = state_.get();
+  switch (state) {
```
Comment (willcrichton) [inline scanner/engine/worker.cpp]: What's the purpose of this switch?

File: scanner/engine/master.cpp [hunk 4]
```diff
@@ -370,40 +438,123 @@ grpc::Status MasterImpl::NextWork(grpc::ServerContext* context,
...
   if (!task_result_.success()) {
+    // Task sampler failed for some reason
     new_work->mutable_io_item()->set_item_id(-1);
     return grpc::Status::OK;
   }
+  new_work->mutable_load_work()->set_job_index(std::get<0>(task_sample_id));
+
+  // Track sample assigned to worker
+  active_task_samples_[node_info->node_id()].insert(task_sample_id);
+  worker_histories_[node_info->node_id()].tasks_assigned += 1;
```
Comment (willcrichton) [inline scanner/engine/master.cpp]: What's the purpose of tasks_assigned? I don't see it used anywhere.

File: scanner/engine/master.cpp [hunk 5]
```diff
@@ -414,28 +565,242 @@ grpc::Status MasterImpl::NewJob(grpc::ServerContext* context,
...
-  job_params_.CopyFrom(*job_params);
+  job_params_.Clear();
+  job_params_.MergeFrom(*job_params);
+  {
+    std::unique_lock<std::mutex> lock(finished_mutex_);
+    finished_ = false;
+  }
+  finished_cv_.notify_one();
+
+  {
```
Comment (willcrichton) [inline scanner/engine/master.cpp]: Is there a difference between `finished_` and `active_job_`?

**Comment by apoms** [comment]
This should be feature complete now!

File: scanner/engine/master.cpp (hunk 5 above)
Comment (apoms) [inline scanner/engine/master.cpp]: Yes, `finished_` indicates all tasks are finished, whereas `active_job_` indicates that the job is still running. A job can be finished_ but still active as it is waiting to clean up workers or write out any final metadata.

File: scanner/engine/master.cpp (hunk 4 above)
Comment (apoms) [inline scanner/engine/master.cpp]: I currently don't use it but I want to use it to keep track of how much worker is assigned to each worker (or even calculate their throughput). I was imagining displaying this on a browser page or as an enhanced progress bar.

File: scanner/engine/worker.cpp (hunk 3 above)
Comment (apoms) [inline scanner/engine/worker.cpp]: Nothing, will remove.

File: python/scannerpy/database.py (hunk 2 above)
Comment (apoms) [inline python/scannerpy/database.py]: Getting rid of it!

File: tests/py_test.py [hunk 6]
```diff
@@ -234,3 +237,217 @@ def test_save_mp4(db):
     f.close()
     table.columns('frame').save_mp4(f.name)
     run(['rm', '-rf', f.name])
+
+@pytest.fixture()
```
Comment (willcrichton) [inline tests/py_test.py]: We should probably start to break out tests into separate files, e.g. `fault_tolerance_test.py` or something.

**Comment by willcrichton** [comment]
Looks good!

**Pull request merged**